File: src/com/za/leetcode/offer/Offer3.java
package com.za.leetcode.offer;

import com.za.leetcode.ListNode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

/**
 * 输入一个链表，按链表值从尾到头的顺序返回一个ArrayList。
 */
public class Offer3 {

    public List<Integer> reverseList(ListNode listNode) {
        if (listNode == null)
            return null;
        List<Integer> result = new ArrayList<>();
        Stack<Integer> tmpStack = new Stack<>();
        while (listNode != null) {
            tmpStack.push(listNode.val);
            listNode = listNode.next;
        }
        while (!tmpStack.empty()) {
            result.add(tmpStack.pop());
        }
        return result;
    }
}

File: src/com/za/leetcode/offer/Offer3.js
function ListNode(x) {
    this.val = x;
    this.next = null;
}

function printListFromTailToHead(head) {
    if (!head)
        return null;
    let array = [];
    while (head) {
        array.unshift(head.val);
        head = head.next;
    }
    return array;
}

function test() {
    let head;
    let h = new ListNode(0);
    head = h;
    for (let i = 1; i < 5; i++) {
        h.next = new ListNode(i);
        h = h.next;
    }
    console.log(printListFromTailToHead(head));
}

test();